File: src/agents/dispatch.ts
import { spawn } from "child_process";
import { createWriteStream, mkdirSync, rmSync, existsSync } from "fs";
import { join } from "path";
import { homedir } from "os";
import { execSync } from "child_process";
import type { AgentName } from "../model/agent.js";
import type { WorkItem } from "../model/work-item.js";
import { AgentStore } from "./agent-store.js";
import { branchName, worktreePath } from "./branch-utils.js";
import { buildClaudePrompt } from "./claude-prompt.js";
import { writeClaudeMd } from "./claude-md.js";
import { AGENTS } from "../model/agent.js";
import { appendEvent } from "../persistence/agent-log.js";

export async function dispatchToAgent(
  agentName: AgentName,
  item: WorkItem,
  repoRoot: string,
  store?: AgentStore,
): Promise<void> {
  const agentStore = store ?? new AgentStore();
  const branch = branchName(agentName, item.id, item.title);
  const wtPath = worktreePath(repoRoot, agentName);

  agentStore.updateAgent(agentName, {
    status: "provisioning",
    workItemId: item.id,
    workItemTitle: item.title,
    workItemSource: item.source,
    branch,
    worktreePath: wtPath,
  });

  appendEvent({
    timestamp: new Date().toISOString(),
    agent: agentName,
    event: "dispatched",
    workItemId: item.id,
    workItemTitle: item.title,
  });

  try {
    // Fetch latest main from origin
    execSync("git fetch origin main", { cwd: repoRoot, stdio: "pipe" });

    // Clean up existing worktree if it exists
    if (existsSync(wtPath)) {
      try {
        execSync(`git worktree remove ${wtPath} --force`, { cwd: repoRoot, stdio: "pipe" });
      } catch {
        rmSync(wtPath, { recursive: true, force: true });
        try {
          execSync("git worktree prune", { cwd: repoRoot, stdio: "pipe" });
        } catch {
          // ignore prune errors
        }
      }
    }

    // Create or reset branch from origin/main
    try {
      execSync(`git branch ${branch} origin/main`, { cwd: repoRoot, stdio: "pipe" });
    } catch {
      execSync(`git branch -f ${branch} origin/main`, { cwd: repoRoot, stdio: "pipe" });
    }

    // Create worktree
    execSync(`git worktree add ${wtPath} ${branch}`, { cwd: repoRoot, stdio: "pipe" });

    // Write CLAUDE.md
    writeClaudeMd(wtPath, AGENTS[agentName].display);

    // npm install in worktree
    execSync("npm install", { cwd: wtPath, stdio: "pipe" });

    // Build prompt
    const prompt = buildClaudePrompt(item, AGENTS[agentName].display);

    // Set up log file
    const logDir = join(homedir(), ".localpipeline", "logs");
    mkdirSync(logDir, { recursive: true });
    const logPath = join(logDir, `agent-${agentName}.log`);
    const logStream = createWriteStream(logPath, { flags: "w" });

    // Spawn claude process
    const child = spawn("claude", ["-p", prompt, "--dangerously-skip-permissions"], {
      cwd: wtPath,
      stdio: ["ignore", "pipe", "pipe"],
    });

    child.stdout.pipe(logStream);
    child.stderr.pipe(logStream);

    // Mark as working with PID
    agentStore.markBusy(agentName, item.id, item.title, item.source, branch, wtPath, child.pid!);

    appendEvent({
      timestamp: new Date().toISOString(),
      agent: agentName,
      event: "working",
      workItemId: item.id,
      workItemTitle: item.title,
    });

    // Handle exit
    child.on("close", (code) => {
      logStream.close();
      if (code === 0) {
        agentStore.markDone(agentName);
        appendEvent({
          timestamp: new Date().toISOString(),
          agent: agentName,
          event: "done",
          workItemId: item.id,
          workItemTitle: item.title,
        });
      } else {
        const message = `Process exited with code ${code}`;
        agentStore.markError(agentName, message);
        appendEvent({
          timestamp: new Date().toISOString(),
          agent: agentName,
          event: "error",
          workItemId: item.id,
          workItemTitle: item.title,
          message,
        });
      }
    });

    child.on("error", (err) => {
      logStream.close();
      agentStore.markError(agentName, err.message);
      appendEvent({
        timestamp: new Date().toISOString(),
        agent: agentName,
        event: "error",
        workItemId: item.id,
        workItemTitle: item.title,
        message: err.message,
      });
    });
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    agentStore.markError(agentName, message);
    appendEvent({
      timestamp: new Date().toISOString(),
      agent: agentName,
      event: "error",
      workItemId: item.id,
      workItemTitle: item.title,
      message,
    });
  }
}

export async function retryAgent(
  agentName: AgentName,
  repoRoot: string,
  store: AgentStore,
): Promise<void> {
  const agent = store.getAgent(agentName);
  if (!agent.workItemId || !agent.workItemTitle || !agent.branch) {
    throw new Error(`Agent ${agentName} missing work item info for retry`);
  }

  const wtPath = worktreePath(repoRoot, agentName);

  // Clean up old worktree
  try {
    execSync(`git worktree remove ${wtPath} --force`, { cwd: repoRoot, stdio: "pipe" });
  } catch {
    // Worktree may already be removed; clean up directory if it lingers
    if (existsSync(wtPath)) {
      rmSync(wtPath, { recursive: true, force: true });
    }
    try {
      execSync("git worktree prune", { cwd: repoRoot, stdio: "pipe" });
    } catch {
      // ignore prune errors
    }
  }

  // Fetch latest main from origin
  execSync("git fetch origin main", { cwd: repoRoot, stdio: "pipe" });

  // Reset the existing branch to origin/main
  try {
    execSync(`git branch -f ${agent.branch} origin/main`, { cwd: repoRoot, stdio: "pipe" });
  } catch {
    // Branch may not exist; create it
    execSync(`git branch ${agent.branch} origin/main`, { cwd: repoRoot, stdio: "pipe" });
  }

  // Re-create worktree
  execSync(`git worktree add ${wtPath} ${agent.branch}`, { cwd: repoRoot, stdio: "pipe" });

  // Reconstruct a WorkItem for dispatch
  const item: WorkItem = {
    id: agent.workItemId,
    title: agent.workItemTitle,
    labels: [],
    source: agent.workItemSource ?? "",
  };

  // Write CLAUDE.md
  writeClaudeMd(wtPath, AGENTS[agentName].display);

  // npm install in worktree
  execSync("npm install", { cwd: wtPath, stdio: "pipe" });

  // Build prompt
  const prompt = buildClaudePrompt(item, AGENTS[agentName].display);

  // Set up log file
  const logDir = join(homedir(), ".localpipeline", "logs");
  mkdirSync(logDir, { recursive: true });
  const logPath = join(logDir, `agent-${agentName}.log`);
  const logStream = createWriteStream(logPath, { flags: "w" });

  // Spawn claude process
  const child = spawn("claude", ["-p", prompt, "--dangerously-skip-permissions"], {
    cwd: wtPath,
    stdio: ["ignore", "pipe", "pipe"],
  });

  child.stdout.pipe(logStream);
  child.stderr.pipe(logStream);

  // Mark as working with PID
  store.markBusy(agentName, item.id, item.title, item.source, agent.branch, wtPath, child.pid!);

  appendEvent({
    timestamp: new Date().toISOString(),
    agent: agentName,
    event: "working",
    workItemId: item.id,
    workItemTitle: item.title,
  });

  // Handle exit
  child.on("close", (code) => {
    logStream.close();
    if (code === 0) {
      store.markDone(agentName);
      appendEvent({
        timestamp: new Date().toISOString(),
        agent: agentName,
        event: "done",
        workItemId: item.id,
        workItemTitle: item.title,
      });
    } else {
      const message = `Process exited with code ${code}`;
      store.markError(agentName, message);
      appendEvent({
        timestamp: new Date().toISOString(),
        agent: agentName,
        event: "error",
        workItemId: item.id,
        workItemTitle: item.title,
        message,
      });
    }
  });

  child.on("error", (err) => {
    logStream.close();
    store.markError(agentName, err.message);
    appendEvent({
      timestamp: new Date().toISOString(),
      agent: agentName,
      event: "error",
      workItemId: item.id,
      workItemTitle: item.title,
      message: err.message,
    });
  });
}

File: src/agents/dispatch.rs
use anyhow::{Context, Result};
use std::path::Path;
use std::process::Stdio;
use tokio::sync::mpsc;

use super::branch::{branch_name, worktree_path};
use super::claude_md::write_claude_md;
use super::claude_prompt::build_prompt;
use super::log::{append_event, new_event};
use super::store::AgentStore;
use crate::app::Action;
use crate::model::agent::AgentName;
use crate::model::work_item::WorkItem;

pub async fn dispatch(
    agent_name: AgentName,
    item: &WorkItem,
    repo_root: &str,
    store: &mut AgentStore,
    action_tx: mpsc::UnboundedSender<Action>,
) -> Result<()> {
    let branch = branch_name(agent_name, &item.id, &item.title);
    let wt_path = worktree_path(repo_root, agent_name);

    // Mark provisioning
    store.mark_provisioning(agent_name, &item.id, &item.title, &branch, &wt_path)?;
    let _ = append_event(&new_event(
        agent_name,
        "dispatched",
        Some(&item.id),
        Some(&item.title),
        None,
    ));

    // Git operations
    run_git(repo_root, &["fetch", "origin", "main"]).await?;

    // Clean up existing worktree
    let wt = Path::new(&wt_path);
    if wt.exists() {
        let _ = run_git(repo_root, &["worktree", "remove", &wt_path, "--force"]).await;
        if wt.exists() {
            tokio::fs::remove_dir_all(&wt_path).await.ok();
        }
    }
    let _ = run_git(repo_root, &["worktree", "prune"]).await;

    // Create branch (force if exists)
    if run_git(repo_root, &["branch", &branch, "origin/main"])
        .await
        .is_err()
    {
        run_git(repo_root, &["branch", "-f", &branch, "origin/main"]).await?;
    }

    // Create worktree
    run_git(repo_root, &["worktree", "add", &wt_path, &branch]).await?;

    // Write CLAUDE.md
    write_claude_md(Path::new(&wt_path), agent_name)?;

    // Build prompt
    let prompt = build_prompt(item, agent_name);

    // Set up log file
    let log_dir = crate::config::data_dir().join("logs");
    std::fs::create_dir_all(&log_dir)?;
    let log_file_path = log_dir.join(format!("agent-{}.log", agent_name.as_str()));
    let log_file = std::fs::File::create(&log_file_path)?;

    // Spawn claude process
    let child = tokio::process::Command::new("claude")
        .args(["-p", &prompt, "--dangerously-skip-permissions"])
        .current_dir(&wt_path)
        .stdin(Stdio::null())
        .stdout(Stdio::from(log_file.try_clone()?))
        .stderr(Stdio::from(log_file))
        .spawn()
        .context("Failed to spawn claude")?;

    let pid = child.id().unwrap_or(0);
    store.mark_working(agent_name, pid)?;
    let _ = append_event(&new_event(
        agent_name,
        "working",
        Some(&item.id),
        Some(&item.title),
        None,
    ));

    // Monitor process in background
    let item_id = item.id.clone();
    let item_title = item.title.clone();
    tokio::spawn(async move {
        let result = child.wait_with_output().await;
        match result {
            Ok(output) if output.status.success() => {
                let _ = append_event(&new_event(
                    agent_name,
                    "done",
                    Some(&item_id),
                    Some(&item_title),
                    None,
                ));
                let _ = action_tx.send(Action::AgentProcessExited(agent_name, true));
            }
            Ok(output) => {
                let msg = format!("Exit code: {}", output.status);
                let _ = append_event(&new_event(
                    agent_name,
                    "error",
                    Some(&item_id),
                    Some(&item_title),
                    Some(&msg),
                ));
                let _ = action_tx.send(Action::AgentProcessExited(agent_name, false));
            }
            Err(e) => {
                let msg = format!("Process error: {e}");
                let _ = append_event(&new_event(
                    agent_name,
                    "error",
                    Some(&item_id),
                    Some(&item_title),
                    Some(&msg),
                ));
                let _ = action_tx.send(Action::AgentProcessExited(agent_name, false));
            }
        }
    });

    Ok(())
}

async fn run_git(cwd: &str, args: &[&str]) -> Result<()> {
    let output = tokio::process::Command::new("git")
        .args(args)
        .current_dir(cwd)
        .output()
        .await
        .with_context(|| format!("Failed to run git {}", args.join(" ")))?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        anyhow::bail!("git {} failed: {}", args.join(" "), stderr);
    }
    Ok(())
}
